### PR TITLE
Fix logic when build scan callback is registered

### PR DESCRIPTION
**Pull request by guylabs**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -54,10 +62,7 @@ bintray {
...
 test {
     environment('TEAMCITY_VERSION', '2019.1')
-}
 
-pluginUnderTestMetadata {
-    // because of weird classloading when testing plugins via TestKit we need to explicitly ensure that the build
-    // scan plugin is on our plugin classpath at test runtime
-    pluginClasspath = pluginClasspath + configurations.compileOnly
+    systemProperty 'buildScanPluginClasspath', configurations.buildScanPlugin.singleFile.absolutePath
+    systemProperty 'gradleEnterprisePluginClasspath', configurations.gradleEnterprisePlugin.singleFile.absolutePath
```
[the remark this answers is not in the document]
Comment (facewindu) [reply build.gradle]: Indeed https://scans.gradle.com/s/2owz6qq4yqdjg/performance/suggestions?open=WyJvdXRPZkRhdGVHcmFkbGVWZXJzaW9uQ2hlY2siLCJub25DYWNoZWFibGVUYXNrT3V0cHV0IiwicmVzb2x1dGlvbkR1cmluZ0NvbmZpZ3VyYXRpb24iXQ#resolution-during-configuration

[the remark this answers is not in the document]
Comment (facewindu) [reply build.gradle]: Fixed https://scans.gradle.com/s/o3txgxvg3jyuq/performance/resolution#summary-configuration-total

File: src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy [hunk 2]
```diff
@@ -51,32 +52,109 @@ class TeamCityBuildScanPluginTest extends Specification {
...
+        when:
+        def result = runner.withArguments("tasks", "-S").build()
+
+        then:
+        outputContainsExpectedMessage(result.output)
+    }
+
+    def "no service messages emitted when build scan or Gradle Enterprise plugin is not applied"() {
+        given:
+        configurePluginClasspath('buildScanPluginClasspath')
```
[the remark this answers is not in the document]
Comment (facewindu) [reply src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy]: done

File: src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy [hunk 3]
```diff
@@ -51,32 +52,109 @@ class TeamCityBuildScanPluginTest extends Specification {
...
+        when:
+        def result = runner.withArguments("tasks", "-S").build()
+
+        then:
+        outputContainsExpectedMessage(result.output)
+    }
+
+    def "service messages emitted for compatible Gradle Enterprise plugin versions"() {
+        given:
+        runner.withGradleVersion("6.0-20191014230037+0000") // Gradle 6.0 snapshot is required for GE plugin 3.0
```
[the remark this answers is not in the document]
Comment (facewindu) [reply src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy]: done

File: src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy [hunk 4]
```diff
@@ -51,32 +52,109 @@ class TeamCityBuildScanPluginTest extends Specification {
...
+
+        then:
+        !outputContainsExpectedMessage(result.output)
+    }
+
+    private void configurePluginClasspath(String classpathSystemPropertyName) {
+        runner.withPluginClasspath(PluginUnderTestMetadataReading.readImplementationClasspath() + [new File(System.getProperty(classpathSystemPropertyName))])
+    }
+
+    private void configureScanPlugin() {
```
[the remark this answers is not in the document]
Comment (facewindu) [reply src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy]: done

File: src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy (hunk 2 above)
[the remark this answers is not in the document]
Comment (facewindu) [reply src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy]: done

File: src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy [hunk 5]
```diff
@@ -51,32 +52,109 @@ class TeamCityBuildScanPluginTest extends Specification {
...
+            
+            apply plugin: "com.gradle.enterprise"
+            gradleEnterprise {
+                buildScan {
+                    server = '${mockScansServer.address}'
+                    publishAlways()
+                }
+            }
+        """.stripIndent()
+        applyPlugin()
```
[the remark this answers is not in the document]
Comment (facewindu) [reply src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy]: done

**Comment by facewindu** [comment]
@etiennestuder all done

File: src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy [hunk 6]
```diff
@@ -115,8 +117,20 @@ class TeamCityBuildScanPluginTest extends Specification {
         !outputContainsExpectedMessage(result.output)
     }
 
-    private void configurePluginClasspath(String classpathSystemPropertyName) {
-        runner.withPluginClasspath(PluginUnderTestMetadataReading.readImplementationClasspath() + [new File(System.getProperty(classpathSystemPropertyName))])
+    def "no service messages emitted when build scan or Gradle Enterprise plugin are not on the classpath"() {
+        given:
+        configurePluginClasspath()
+        applyPlugin()
```
Comment (etiennestuder) [inline src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy]: @facewindu Does this test implementation really match the description of the test? If so, why ar we configuring the plugin classpath and applying the plugin?

File: src/main/java/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPlugin.java [hunk 7]
```diff
@@ -22,23 +24,36 @@ public void apply(Project project) {
...
+            ExtensionContainer extensions = project.getExtensions();
+            try {
+                Class<?> buildScanExtensionClass = Class.forName(BUILD_SCAN_EXTENSION);
+                if (extensions.findByType(buildScanExtensionClass) != null) {
+                    registerCallback(project);
+                    return;
+                }
+            } catch (ClassNotFoundException ignore) {
             }
+            project.getPluginManager().withPlugin(BUILD_SCAN_PLUGIN_ID, appliedPlugin -> registerCallback(project));
```
Comment (etiennestuder) [inline src/main/java/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPlugin.java]: @guylabs @facewindu I don't understand this logic above where we either find the build scan extension or use the plugin manager? Can you please clarify?

File: src/main/java/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPlugin.java [hunk 8]
```diff
@@ -22,23 +24,36 @@ public void apply(Project project) {
...
         }
     }
 
-    private static boolean supportsScanPublishedListener(@Nullable BuildScanExtension extension) {
-        if (extension == null) return false;
-        for (Method method : extension.getClass().getMethods()) {
+    private static void registerCallback(Project project) {
+        BuildScanExtension buildScanExtension = project.getExtensions().getByType(BuildScanExtension.class);
```
Comment (etiennestuder) [inline src/main/java/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPlugin.java]: We could get to this line even with `Class.forName(BUILD_SCAN_EXTENSION)` having thrown an exception.

File: src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy [hunk 9]
```diff
@@ -51,32 +54,121 @@ class TeamCityBuildScanPluginTest extends Specification {
...
+        when:
+        def result = runner.withArguments("tasks", "-S").build()
+
+        then:
+        outputContainsExpectedMessage(result.output)
+    }
+
+    def "service messages emitted for compatible Gradle Enterprise plugin versions"() {
+        given:
+        runner.withGradleVersion("6.0-rc-1") // Gradle 6.0 is required for GE plugin 3.0
```
Comment (etiennestuder) [inline src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy]: @facewindu This can be updated to Gradle 6.0 now.

File: src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy [hunk 10]
```diff
@@ -51,32 +54,121 @@ class TeamCityBuildScanPluginTest extends Specification {
...
+            gradleEnterprise {
+                buildScan {
+                    server = '${mockScansServer.address}'
+                    publishAlways()
+                }
+            }
+        """.stripIndent()
+    }
+
+    private void applyPlugin() {
```
Comment (etiennestuder) [inline src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy]: This should be renamed to something like `applyGradleBuildScanTeamCityPlugin`.

File: src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy [hunk 11]
```diff
@@ -51,32 +54,121 @@ class TeamCityBuildScanPluginTest extends Specification {
...
+
+        then:
+        !outputContainsExpectedMessage(result.output)
+    }
+
+    private void configureBuildScanPlugin(String classpathSystemPropertyName = null) {
+        runner.withPluginClasspath(PluginUnderTestMetadataReading.readImplementationClasspath() + (classpathSystemPropertyName ? [new File(System.getProperty(classpathSystemPropertyName))] : []))
+    }
+
+    private void configureScanPlugin() {
```
Comment (etiennestuder) [inline src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy]: rename to configureBuildScanPlugin

File: src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy [hunk 12]
```diff
@@ -51,32 +54,121 @@ class TeamCityBuildScanPluginTest extends Specification {
...
+        applyPlugin()
+
+        when:
+        def result = runner.withArguments("tasks", "-S").build()
+
+        then:
+        !outputContainsExpectedMessage(result.output)
+    }
+
+    private void configureBuildScanPlugin(String classpathSystemPropertyName = null) {
```
Comment (etiennestuder) [inline src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy]: Rename to `configureTestKitRunnerPluginClasspath`.

File: src/main/java/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPlugin.java (hunk 7 above)
Comment (guylabs) [reply src/main/java/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPlugin.java]: As the Gradle Enterprise plugin is a settings plugin and is not applied to a `Project` itself, the `pluginManager.withPlugin()` callback would never be called. Therefore we check for the existence of the extension to see if it is already applied via the settings script. If true, we register the callback, if not, we use the  `pluginManager.withPlugin()` callback for the Build Scan plugin, as this is registered on the `Project` level.

Depending when the gradle-build-scan-teamcity-plugin is applied/evaluated one of the above paths will be taken, as the extension class type is the same on both plugins. If we know that the gradle-build-scan-teamcity-plugin is always applied after the Gradle Enterprise or Build Scan plugin, we could get rid of the `pluginManager.withPlugin()` callback, as the extension will always be existent. We kept both to make the application of the gradle-build-scan-teamcity-plugin not order dependent.

Comment (etiennestuder) [reply src/main/java/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPlugin.java]: What is your assumption that the build scan plugin or GE plugin is applied at all? The scenario of the scan/GE plugin not being applied is still not covered, and it's what I have pointed out repeatedly in the tests.

Als just in terms of code style, the current flow is ver hard to follow and needs to be improved.

File: src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy (hunk 6 above)
Comment (facewindu) [reply src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy]: I reworked it to be clearer. We were adding the TC plugin to the classpath, but not the GE/build scan plugin (as no parameters were given to configurePluginClasspath() )

File: src/main/java/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPlugin.java (hunk 7 above)
Comment (facewindu) [reply src/main/java/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPlugin.java]: I reworked it a little bit, but the essence stays the same as Guy explained.
I dont find a clever way to achieve all of that.

File: src/main/java/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPlugin.java (hunk 8 above)
Comment (facewindu) [reply src/main/java/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPlugin.java]: reworked, but this was not a problem

File: src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy (hunk 9 above)
Comment (facewindu) [reply src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy]: done

File: src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy (hunk 10 above)
Comment (facewindu) [reply src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy]: done

File: src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy (hunk 11 above)
Comment (facewindu) [reply src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy]: done

File: src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy (hunk 12 above)
Comment (facewindu) [reply src/test/groovy/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPluginTest.groovy]: done

**Comment by etiennestuder** [comment]
@facewindu @guylabs I'm going to close this PR. I've taken parts of the test fixtures and reimplemented the plugin production cod.
